### PR TITLE
[Loads] Check context instruction for context-sensitive derefability

### LINKED ISSUE #108854: isSafeToSpeculativelyExecute seems not safe to use from SimplifyCFG
The `llvm::isSafeToSpeculativelyExecute` function seems to be only safe to use when not speculating the result of moving instructions, since it may consider attributes that will get discarded during moving it later (e.g. dereferenceable). However, it can be seen that `SimplifyCFG:: dominatesMergePoint` tries to use it exactly in that way. Otherwise, like seen in the small snippet below, that can allow violating the semantics of attributes of the IR which, won't be valid after speculating is done (which is expected to discard those attributes).

The invalid speculation appears to happen here: https://github.com/llvm/llvm-project/blob/b592917eec407a01ef07334af5aafe6c8922c80a/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L461-L465

This can be observed by running `opt -passes=simplifycfg this.ll` and observing that it returns IR with only basic block, which violates the semantics of `%4` being a conditional load in the original IR specification (%3 is not dereferencable after hoisting for `foldTwoEntryPHINode`):

```llvm
define i64 @julia_collect_81(i1 %0, ptr nonnull align 8 dereferenceable(8) %1) {
top:
  br i1 %0, label %L82, label %L85

L82:
  %2 = load ptr, ptr %1, align 8, !dereferenceable !1, !align !1
  %3 = load i64, ptr %2, align 8
  br label %L85

L85:
  %4 = phi i64 [ %3, %L82 ], [ 0, %top ]
  ret i64 %4
}

!0 = !{}
!1 = !{i64 8}
```

What seems the best course of action to fix this pass mis-speculation? I think LICM and LoopRotationUtils and SpeculativeExecution may share same issue ~~but SpeculativeExecution has a private `AllPrecedingUsesFromBlockHoisted` check that it appears to have been using to avoid this issue (present since that pass was introduced in 154eb5aa1d9c829715d13c9a047d4498be4d2e2a)~~

### PR DESCRIPTION
**Pull request by nikic**
If a dereferenceability fact is provided through `!dereferenceable` (or similar), it may only hold on the given control flow path. When we use `isSafeToSpeculativelyExecute()` to check multiple instructions, we might make use of `!dereferenceable` information that does not hold at the speculation target. This doesn't happen when speculating instructions one by one, because `!dereferenceable` will be dropped while speculating.

Fix this by checking whether the instruction with `!dereferenceable` dominates the context instruction. If this is not the case, it means we are speculating, and cannot guarantee that it holds at the speculation target.

Fixes https://github.com/llvm/llvm-project/issues/108854.

### CODE REVIEW

File: llvm/lib/Analysis/Loads.cpp [hunk 1]
```diff
@@ -104,6 +104,17 @@ static bool isDereferenceableAndAlignedPointer(
...
       return false;
+    // When using something like !dereferenceable on a load, the
+    // dereferenceability may only be valid on a specific control-flow path.
+    // If the instruction doesn't dominate the context instruction, we're
+    // asking about dereferenceability under the assumption that the
+    // instruction has been speculated to the point of the context instruction,
+    // in which case we don't know if the dereferenceability info still holds.
+    // We don't bother handling allocas here, as they aren't speculatable
+    // anyway.
+    auto *I = dyn_cast<Instruction>(V);
```
Comment (tschuett) [inline llvm/lib/Analysis/Loads.cpp:115]:
```
if (auto *I = dyn_cast<Instruction>(V))
  if (!isa<AllocaInst>(I))
     return CtxI && isValidAssumeForContext(I, CtxI, DT);
```

**Comment by efriedma-quic** [comment]
Please update the documentation for isSafeToSpeculativelyExecute() to specify the semantics in the case where the operands of the instruction don't dominate CtxI.

**Comment by nikic** [comment]
I added some wording to isSafeToSpeculativelyExecute(), but not sure if this is what you had in mind.

**Comment by efriedma-quic** [comment]
Yes, that makes sense, thanks.

**Review by dtcxzyw — APPROVED** [review]
Nice catch!

**Comment by danilaml** [comment]
Is it even possible for `isSafeToSpeculativelyExecute` to return `true` now with the default (nullptr) CtxI?

**Comment by nikic** [comment]
> Is it even possible for `isSafeToSpeculativelyExecute` to return `true` now with the default (nullptr) CtxI for loads? Can `isDereferenceableAndAlignedPointer` just short-circuit to `false` with null CtxI?

Yes, it's possible for anything where the derefability is not context-sensitive (like globals, dereferenceable arguments, allocas, etc).

**Comment by danilaml** [comment]
@nikic I mean not in theory but currently. I don't see those (except allocas) handled anywhere unless I'm missing something?

**Comment by nikic** [comment]
> @nikic I mean not in theory but currently. I don't see those (except allocas) handled anywhere unless I'm missing something?

The `I &&` part handles those. If it's a global or argument (thus not an instruction) we'll fall through to the return true.

**Comment by danilaml** [comment]
@nikic I see. Thank. Probably not gonna help me come up with a reproducer for a weird behavior/bug I was seeing with LoopVectorizer where it produced loads from poison pointers when it thought it didn't need to predicate a load (after this patch it always needs to since loads are never safe to speculate in its check but I think it's just masking whatever is going on).

**Comment by danilaml** [comment]
@nikic By the way, are there plans to support allocation functions other than alloca in this check? I don't see currently any llvm passes assigning dereferenceable(_or_null) attribute to something like `malloc(42)` , but I don't see why not and in that case this should also be something not reliant on the context.

**Comment by nikic** [comment]
> @nikic By the way, are there plans to support allocation functions other than alloca in this check? I don't see currently any llvm passes assigning dereferenceable(_or_null) attribute to something like `malloc(42)` , but I don't see why not and in that case this should also be something not reliant on the context.

I believe we don't mark allocation return values as dereferencable because it would imply a too strong property right now (staying dereferenceable even after the allocation was freed).

**Comment by danilaml** [comment]
@nikic ah, yeah. I remember the issue with dereferenceable w.r.t. free discussion. Guess will have to roll something for allocations that don't have this problem.

**Comment by danilaml** [comment]
@nikic one thing I've discovered with this fact is that now `load` is not dereferenceable at "load" itself due to `isValidAssumeForContext` quirk of not allowing `assume` affect itself. Not sure if it's intentional for attributes here - makes it look like it's not safe to speculate load in its current position. There is also some friction with instruction vs iterator (i.e. when you only have insertion point at the end of bb so can't easily turn it into an instruction), but that's a different matter.

**Comment by nikic** [comment]
@danilaml We could pass AllowEphemerals=true to isValidAssumeForContext.

**Comment by danilaml** [comment]
@nikic Does `AllowEphemerals` actually do what we want in that case? Haven't looked in detail, but on the surface it seemed way more involved than just `if (I == CtxI) return true;`.

**Comment by danilaml** [comment]
@nikic also, another situation the current impl doesn't seem to account for is IR like
```llvm
bb1:
  %p1 = load dereferenceable(N)
...
bb2:
  %p2 = load dereferenceable(M)
...
common:
  %p = phi ptr [%p1, %bb1], [%p2, %bb2]
  %res = load i32 %p ; <-- dereferenceable(min(N,M)) at least
```

**Pull request merged**